R/tostring.R

```diff
@@ -805,14 +1087,159 @@ wrap_string <- function(str, width, collapse = NULL) {
 }
 
 .go_stri_wrap <- function(str, w) {
+  if (is.na(w) || is.nan(w)) {
+    browser()
+  }
+  if (w < 1) {
+    return(str)
+  }
   stringi::stri_wrap(str,
     width = w,
     normalize = FALSE, # keeps spaces
     simplify = TRUE, # If FALSE makes it a list with str elements
-    indent = 0
+    indent = 0,
+    use_length = FALSE # incase the defaul changes, use actual char widths
   )
 }
 
+#' @rdname wrap_string_ttype
+#' @export
+split_word_ttype <- function(str, width, fontspec, min_ok_chars) {
+  chrs <- strsplit(str, "")[[1]]
+  nctt_chars <- nchar_ttype(chrs, fontspec, raw = TRUE)
+  ok <- which(cumsum(nctt_chars) <= width)
+  if (length(ok) < min_ok_chars || length(chrs) - length(ok) < min_ok_chars) {
+    list(
+      ok = character(),
+      remainder = str
+    )
+  } else {
+    list(
+      ok = substr(str, 1, length(ok)),
+      remainder = substr(str, length(ok) + 1, nchar(str))
+    )
+  }
+}
+
+## need a separate path here because **the number of characters**
+## in each part is no longer going to be constant the way it
+## was for monospace
+## this is much slower but still shouldn't be a bottleneck, if it is we'll
+## have to do something else
+#' wrap string given a Truetype font
+#'
+#' @inheritParams wrap_string
+#' @param min_ok_chars numeric(1). Number of minimum characters which much remain
+#' on either side when a word is split
+#' @param wordbreak_ok logical(1). Should breaking within a word be allowed. If, `FALSE`,
+#' attempts to wrap a string to a width narrower than its widest word will result
+#' in an error.
+#' @return `str`, broken up into a word-wrapped vector
+#' @export
+wrap_string_ttype <- function(str,
+                              width,
+                              fontspec,
+                              collapse = NULL,
+                              min_ok_chars = min(floor(nchar(str) / 2), 4, floor(width / 2)),
+                              wordbreak_ok = TRUE) {
+  newdev <- open_font_dev(fontspec)
+  if (newdev) {
+    on.exit(close_font_dev())
+  }
+
+  rawspls <- strsplit(str, "[[:space:]](?=[^[:space:]])", perl = TRUE)[[1]] # preserve all but one space
```

Comment:
It is of course fine but why do you enforce it? stringi should be able to handle multiple spaces correctly

Reply:
strsplit is a base R function, its not from stringi. I wouldn't be surprised if stringi has something similar, but this seemed fine.

In terms of why we care about multiples, we (arguably) want

`"hello   there"` (3 spaces) to wrap to `"hello  ", "there"`  if the width is 7, not  `"hello", "there"`. This is a weird corner case that we could easily say isn't supported at all (ie no putting extra space in labels to fiddle with wrapping behavior) but someday people are gonna want to fiddle with wrapping behavior...

Reply:
but "preserve only one space" seems to indicate that the result is `"hello ", "there"`, hence removing the other spaces. If it keeps them, I would rephrase the comment for clarity. Again it is not super relevant to me, as you wish ^^. For context, we went towards supporting this fiddling at the end by using stringi as people were, as you said, asking to do weird corner cases

Reply:
its preserving *all but one* space, not only one space. I don't really feel that strongly about how the comment is phrased though

Reply:
I see, It makes sense now! it is fine like it is ^^